### ISSUE TITLE
🐛 Bug Report: Catalog cannot process templates with deeply nested parameters

### ISSUE THREAD

**Issue by axiac** [backstage/backstage#23295]
### 📜 Description

The templates having parameters with deeply nested structure cannot be stored in the catalog.
The ingestion fails with message `Failed to stitch template:default/abcd, error: insert into "search" ("entity_id", "key", "original_value", "value") values ($1, $2, $3, $4), ($5, $6, $7, $8), ... ($197, $198, $199, $200) - value too long for type character varying(255)`

## Investigation

The insertion in the `search` table is done on [the last statement of function `performStitching()`](https://github.com/backstage/backstage/blob/master/plugins/catalog-backend/src/database/operations/stitcher/performStitching.ts/#L242) and the values to be inserted are computed by [function `buildEntitySearch()`](https://github.com/backstage/backstage/blob/master/plugins/catalog-backend/src/database/operations/stitcher/buildEntitySearch.ts#L175-L221). This function extracts from entity the data that it wants to save, then calls [function `mapToRows()`](https://github.com/backstage/backstage/blob/master/plugins/catalog-backend/src/database/operations/stitcher/buildEntitySearch.ts#L131-L166) that produces the final records to be persisted.

The function `mapToRows()` receives a list of (`key`, `value`) pairs as argument and produces a list of search rows (with four columns) to be persisted. Closer examination of its code reveals that it **ignores the pairs** whose [property `key` is longer than a certain threshold](https://github.com/backstage/backstage/blob/master/plugins/catalog-backend/src/database/operations/stitcher/buildEntitySearch.ts#L145-L161) and it uses `null` as `value` when [`key` is short enough but `value` is too long](https://github.com/backstage/backstage/blob/master/plugins/catalog-backend/src/database/operations/stitcher/buildEntitySearch.ts#L153-L159) (there is a threshold for `value`, too).
However, it uses these safety measures [only when `value` is neither `null`, nor `undefined`](https://github.com/backstage/backstage/blob/master/plugins/catalog-backend/src/database/operations/stitcher/buildEntitySearch.ts#L136-L143). When `value` is `null` or `undefined`, it does not do any check on the length of `key` and it ends up with the SQL error listed above.

## Discussion

As revealed in [this comment](https://github.com/backstage/backstage/issues/21863#issuecomment-1952218811) on issue #21863, changing the column types to `TEXT`, to be able to accommodate longer values, does not seem to be a solution.
Issue #21863 describes the same underlying database issue in a different table.

Searching the PRs and issues in GitHub revealed that function `mapToRows()` has been changed to [skip the entries with large keys](https://github.com/backstage/backstage/pull/5684/files#diff-919fe7004a5cfcb6c4568e0363f5c139d935beb5bc6309e7efdb6adb629ce6a3R140) in PR #5684, but only when the values are not `null` or `undefined`.

I think that a good solution for this issue is to apply the same validation and ignore the keys that are too long also when the value is `null` or `undefined` but this is just a guess based on the existing code of function `mapToRows()`.

What do you think?

### 👍 Expected behavior

It should save the template in the catalog, no matter the depth of the structure of their parameters.

### 👎 Actual Behavior with Screenshots

The template is not saved in the catalog.
It also does not show up either as "pending" or as "failed" in the DevTools unprocessed entities page.

### 👟 Reproduction steps

Below is an example template adjusted to reveal the bug. It does not do anything, maybe it does not even makes sense. The real use case is much larger and uses multiple levels of nested objects with `dependency` and `oneOf` and whatnot.

What is important here is that it cannot be saved using PostgreSQL but it can be saved using SQLite (which kindof make it a valid entity for ingestion, even if it does not make much sense as template.)

```yaml
apiVersion: scaffolder.backstage.io/v1beta3
kind: Template
metadata:
  name: abcd
  title: Template with long key paths in parameters
  description: Bug demo
spec:
  owner: backstage/techdocs-core
  type: service
  parameters:
    - title: Fill in some steps
      properties:
        fullName_1234567890_abcdefghij_1234567890_abcdefghij_1234567890_abcdefghij_1234567890_abcdefghij_1234567890_abcdefghij_1234567890_abcdefghij:
          type: object
          title: Full name
          properties:
            firstName_1234567890_abcdefghij_1234567890_abcdefghij_1234567890_abcdefghij_1234567890_abcdefghij_1234567890_abcdefghij_1234567890_abcdefghij:
              type: string
              title: First name
              default: null                       # crash, boom, bang!
            lastName_1234567890_abcdefghij_1234567890_abcdefghij_1234567890_abcdefghij_1234567890_abcdefghij_1234567890_abcdefghij_1234567890_abcdefghij:
              type: string
              title: Last name
  steps: []
  output:
    links: []
```

### 🖥️ Your Environment

```
OS:   Darwin 21.6.0 - darwin/arm64
node: v18.19.0
yarn: 1.22.21
cli:  0.25.1 (installed)
backstage:  1.22.1

Dependencies:
  @backstage/app-defaults                                          1.4.7
  @backstage/backend-app-api                                       0.5.10
  @backstage/backend-common                                        0.19.9, 0.20.1
  @backstage/backend-defaults                                      0.2.9
  @backstage/backend-dev-utils                                     0.1.3
  @backstage/backend-openapi-utils                                 0.1.2
  @backstage/backend-plugin-api                                    0.6.9
  @backstage/backend-tasks                                         0.5.14
  @backstage/backend-test-utils                                    0.2.10
  @backstage/catalog-client                                        1.5.2
  @backstage/catalog-model                                         1.4.3
  @backstage/cli-common                                            0.1.13
  @backstage/cli-node                                              0.2.2
  @backstage/cli                                                   0.25.1
  @backstage/config-loader                                         1.6.1
  @backstage/config                                                1.1.1
  @backstage/core-app-api                                          1.11.3
  @backstage/core-compat-api                                       0.1.1
  @backstage/core-components                                       0.13.10
  @backstage/core-plugin-api                                       1.8.2
  @backstage/dev-utils                                             1.0.26
  @backstage/errors                                                1.2.3
  @backstage/eslint-plugin                                         0.1.4
  @backstage/frontend-plugin-api                                   0.5.0
  @backstage/integration-aws-node                                  0.1.8
  @backstage/integration-react                                     1.1.23
  @backstage/integration                                           1.8.0
  @backstage/plugin-api-docs                                       0.10.3
  @backstage/plugin-app-backend                                    0.3.57
  @backstage/plugin-app-node                                       0.1.9
  @backstage/plugin-auth-backend-module-atlassian-provider         0.1.1
  @backstage/plugin-auth-backend-module-gcp-iap-provider           0.2.3
  @backstage/plugin-auth-backend-module-github-provider            0.1.6
  @backstage/plugin-auth-backend-module-gitlab-provider            0.1.6
  @backstage/plugin-auth-backend-module-google-provider            0.1.6
  @backstage/plugin-auth-backend-module-oauth2-provider            0.1.6
  @backstage/plugin-auth-backend-module-oauth2-proxy-provider      0.1.1
  @backstage/plugin-auth-backend-module-okta-provider              0.0.2
  @backstage/plugin-auth-backend                                   0.20.3
  @backstage/plugin-auth-node                                      0.4.3
  @backstage/plugin-catalog-backend-module-incremental-ingestion   0.4.13
  @backstage/plugin-catalog-backend-module-scaffolder-entity-model 0.1.6
  @backstage/plugin-catalog-backend-module-unprocessed             0.3.6
  @backstage/plugin-catalog-backend                                1.16.1
  @backstage/plugin-catalog-common                                 1.0.20
  @backstage/plugin-catalog-import                                 0.10.5
  @backstage/plugin-catalog-node                                   1.6.1
  @backstage/plugin-catalog-react                                  1.9.3
  @backstage/plugin-catalog-unprocessed-entities                   0.1.7
  @backstage/plugin-catalog                                        1.16.1
  @backstage/plugin-cost-insights-common                           0.1.2
  @backstage/plugin-cost-insights                                  0.12.18
  @backstage/plugin-devtools-backend                               0.2.6
  @backstage/plugin-devtools-common                                0.1.8
  @backstage/plugin-devtools                                       0.1.8
  @backstage/plugin-events-node                                    0.2.18
  @backstage/plugin-github-actions                                 0.6.10
  @backstage/plugin-home-react                                     0.1.7
  @backstage/plugin-home                                           0.6.1
  @backstage/plugin-org                                            0.6.19
  @backstage/plugin-permission-backend                             0.5.32
  @backstage/plugin-permission-common                              0.7.12
  @backstage/plugin-permission-node                                0.7.20
  @backstage/plugin-permission-react                               0.4.19
  @backstage/plugin-proxy-backend                                  0.4.7
  @backstage/plugin-scaffolder-backend-module-azure                0.1.1
  @backstage/plugin-scaffolder-backend-module-bitbucket            0.1.1
  @backstage/plugin-scaffolder-backend-module-gerrit               0.1.1
  @backstage/plugin-scaffolder-backend-module-github               0.1.1
  @backstage/plugin-scaffolder-backend-module-gitlab               0.2.12
  @backstage/plugin-scaffolder-backend                             1.20.0
  @backstage/plugin-scaffolder-common                              1.4.5
  @backstage/plugin-scaffolder-node                                0.2.10
  @backstage/plugin-scaffolder-react                               1.7.1
  @backstage/plugin-scaffolder                                     1.17.1
  @backstage/plugin-search-backend-module-catalog                  0.1.13
  @backstage/plugin-search-backend-module-pg                       0.5.18
  @backstage/plugin-search-backend-module-techdocs                 0.1.13
  @backstage/plugin-search-backend-node                            1.2.13
  @backstage/plugin-search-backend                                 1.4.9
  @backstage/plugin-search-common                                  1.2.10
  @backstage/plugin-search-react                                   1.7.5
  @backstage/plugin-search                                         1.4.5
  @backstage/plugin-techdocs-backend                               1.9.2
  @backstage/plugin-techdocs-module-addons-contrib                 1.1.4
  @backstage/plugin-techdocs-node                                  1.11.1
  @backstage/plugin-techdocs-react                                 1.1.15
  @backstage/plugin-techdocs                                       1.9.3
  @backstage/plugin-user-settings                                  0.8.0
  @backstage/release-manifests                                     0.0.11
  @backstage/test-utils                                            1.4.7
  @backstage/theme                                                 0.5.0
  @backstage/types                                                 1.1.1
  @backstage/version-bridge                                        1.0.7
```

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

Yes I am willing to submit a PR!

**Comment by freben** [comment]
Agreed! Thanks for the thorough analysis, that's perfectly correct. Applying the length limitation farther up so that it applies in all branches is correct. Go ahead and make a PR if you have the opportunity! 🙏

**Comment by axiac** [comment]
This has been fixed in #23337 and released in Backstage v1.24
